src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs

```diff
@@ -100,5 +106,70 @@ public static class TagHelperOutputExtensions
                 tagHelperOutput.Attributes.Remove(attribute);
             }
         }
+
+        private static void CopyHtmlAttributeMaintainOrder(
+            int allAttributeIndex,
+            TagHelperOutput tagHelperOutput,
+            TagHelperContext context)
+        {
+            var existingAttribute = context.AllAttributes[allAttributeIndex];
+            var copiedAttribute = new TagHelperAttribute
+            {
+                Name = existingAttribute.Name,
+                Value = existingAttribute.Value,
+                Minimized = existingAttribute.Minimized
+            };
+
+            for (var i = allAttributeIndex - 1; i >= 0; i--)
+            {
+                // Move backwards through context.AllAttributes from the provided index until we find a familiar
+                // attribute in tagHelperOutput where we can insert the copied value in front of.
+                var previousAllAttributeName = context.AllAttributes[i].Name;
+                if (TryFindNameAndInsertAttribute(
+                    previousAllAttributeName,
+                    insertOffset: 1,
+                    attributeToInsert: copiedAttribute,
+                    attributes: tagHelperOutput.Attributes))
+                {
+                    return;
+                }
+            }
+
+            for (var i = allAttributeIndex + 1; i < context.AllAttributes.Count; i++)
+            {
+                // Move forward through context.AllAttributes from the provided index until we find a familiar
+                // attribute in tagHelperOutput where we can insert the copied value.
+                var nextAllAttributeName = context.AllAttributes[i].Name;
+                if (TryFindNameAndInsertAttribute(
+                    nextAllAttributeName,
+                    insertOffset: 0,
+                    attributeToInsert: copiedAttribute,
+                    attributes: tagHelperOutput.Attributes))
+                {
+                    return;
+                }
+            }
+
+            // Couldn't determine the attribute's location, add it to the end.
+            tagHelperOutput.Attributes.Add(copiedAttribute);
+        }
+
+        private static bool TryFindNameAndInsertAttribute(
+            string name,
+            int insertOffset,
+            TagHelperAttribute attributeToInsert,
+            TagHelperAttributeList attributes)
+        {
+            for (var j = 0; j < attributes.Count; j++)
+            {
+                if (string.Equals(name, attributes[j].Name, StringComparison.OrdinalIgnoreCase))
+                {
+                    attributes.Insert(j + insertOffset, attributeToInsert);
```

Comment:
This results in an odd mix of responsibilities. Instead do the insertion in `CopyHtmlAttribute()` and change this method to
``` c#
private static int FindFirstMatch(TagHelperAttributeList attributeList, string name)
{
    for (var index = 0; index < attributes.Count; index++)
    {
        if (string.Equals(name, attributeList[index].Name, StringComparison.OrdinalIgnoreCase))
        {
            return index;
        }
    }

    return -1;
}
```